node/src/components/consensus/highway_core/active_validator.rs

```diff
@@ -124,47 +122,70 @@ impl<C: Context> ActiveValidator<C> {
             next_round_exp: state.params().init_round_exp(),
             next_timer: state.params().start_timestamp(),
             next_proposal: None,
-            unit_hash_file,
+            unit_file,
             own_last_unit,
             target_ftt,
             paused: false,
         };
-        let effects = av.schedule_timer(start_time, state);
+        let mut effects = av.schedule_timer(start_time, state);
+        effects.push(av.send_ping(current_time));
+        effects.extend(
+            av.own_last_unit
+                .as_ref()
+                .map(|unit| Effect::NewVertex(ValidVertex(Vertex::Unit(unit.clone()))))
+                .into_iter(),
+        );
         (av, effects)
     }
 
-    fn read_last_unit<P: AsRef<Path>>(path: P) -> io::Result<C::Hash> {
+    fn read_last_unit<P: AsRef<Path>>(path: P) -> io::Result<SignedWireUnit<C>> {
         let mut file = File::open(path)?;
         let mut bytes = Vec::new();
         file.read_to_end(&mut bytes)?;
         Ok(serde_json::from_slice(&bytes)?)
     }
 
-    fn write_last_unit(&mut self, hash: C::Hash) -> io::Result<()> {
-        // If there is no unit_hash_file set, do not write to it
-        let unit_hash_file = if let Some(file) = self.unit_hash_file.as_ref() {
+    fn write_last_unit(&mut self, swunit: SignedWireUnit<C>) -> io::Result<()> {
+        // If there is no unit_file set, do not write to it
+        let unit_file = if let Some(file) = self.unit_file.as_ref() {
             file
         } else {
             return Ok(());
         };
 
-        // Otherwise, set own_last_unit to the specified hash
-        self.own_last_unit = Some(hash);
-
         // Create the file (and its parents) as necessary
-        if let Some(parent_directory) = unit_hash_file.parent() {
+        if let Some(parent_directory) = unit_file.parent() {
             fs::create_dir_all(parent_directory)?;
         }
-        let mut file = File::create(unit_hash_file)?;
+        let mut file = File::create(unit_file)?;
 
         // Finally, write the data to file we created
-        let bytes = serde_json::to_vec(&hash)?;
+        let bytes = serde_json::to_vec(&swunit)?;
+
         file.write_all(&bytes)
     }
 
+    /// Cleans up the validator disk state.
+    /// Deletes all unit files.
+    pub(crate) fn cleanup(&self) -> io::Result<()> {
+        let unit_file = if let Some(file) = self.unit_file.as_ref() {
+            file
+        } else {
+            return Ok(());
+        };
+
+        fs::remove_file(unit_file)
+    }
+
     fn can_vote(&self, state: &State<C>) -> bool {
-        self.own_last_unit
-            .map_or(true, |ref hash| state.has_unit(hash))
+        self.own_last_unit.as_ref().map_or(true, |swunit| {
+            state.has_unit(&swunit.hash())
+                && swunit
+                    .wire_unit()
+                    .panorama
+                    .iter_correct_hashes()
+                    .all(|hash| state.has_unit(hash))
```

Comment:
That's unnecessary: It is implied by `state.has_unit(&swunit.hash())`.

Reply:
Right, my initial version was only checking for the unit's panorama, not the unit itself.

Reply:
Fixed in [`4289f2d` (#1134)](https://github.com/CasperLabs/casper-node/pull/1134/commits/4289f2d02aef82fd23265003ad14867761eafbf9)